### ISSUE TITLE
[Feature Request] Disable Undo Point Option

### ISSUE THREAD

**Issue by gentlegiantJGC** [Amulet-Team/Amulet-Core#255]
# The Problem
The undo point operation takes quite a while to run.
Many users have expressed wanting to be able to disable this.
The undo point was added in as a mechanism to allow users to undo changes but also as a mechanism of freeing RAM.

# Chunk Storage
## About
When a chunk is loaded, Amulet needs to hold a reference to it so that if it is requested again the same chunk object can be returned. From now on we will refer to this as the RAM database.

There needs to be some condition when chunks can be unloaded from the RAM database to free memory, otherwise memory usage will grow over time.
Chunks cannot be unloaded if they have changed otherwise those changes would be lost.

Currently this condition is a manual function that unloads all chunks.
It would be better if chunks were automatically unloaded when they are no longer referenced (and are not changed)

The history database was a mechanism to allow the chunk to be persisted to disk for later loading and unloaded from the RAM database to free up RAM.
When loading a chunk the history database is first checked and if it contains the chunk it is loaded. If not it is loaded from the raw level.

One hurdle to removing the undo point is the behaviour when a chunk is changed but not marked as changed. Currently these changes are discarded but without a backup to revert to those changes will remain.
Should this be the correct behaviour or should a solution be found? Perhaps caching the previous state in RAM.

Changing the undo point behaviour is possible but it is currently a rather core part of how Amulet operates so it can't currently just be disabled.

To remove the undo point we would need to disable the disk caching system and all changed chunks would have to exist purely in RAM. We would also need to make sure that the chunks in RAM were not cleared.

## History Database on disk.
Having a history database on disk allows RAM to be freed so is the best solution for most users with limited RAM.

### History Database in RAM.
This is the fastest way to have a history system but requires a lot of RAM

## No History Database.
This is the fastest solution but requires a bit of RAM.
Changed chunks cannot be unloaded from the RAM database.
Will have to manage the case where an operation changes a chunk but does not mark it as changed.

## Prerequisites
Amulet-Team/Amulet-Map-Editor#687 Chunk management improvement
Amulet-Team/Amulet-Map-Editor#688 Chunk ownership
- Add a system to create the undo point during the operation rather than at the end. This would allow unloading changed chunks.

If I have more thoughts on this I will modify this.

**Comment by gentlegiantJGC** [comment]
I have been thinking this over in relation to a couple of other issues to find the best solution.
The changes I want to make are rather substantial so would be part of a 2.0 release. Amulet V0 will continue using Core V1 and the Qt rewrite (Amulet V1) will use the new version of the library.

The changes that I want to make mean that there won't be a distinct undo creation stage.

**Comment by gentlegiantJGC** [comment]
I have thought about this some more and I think I have a solution.
I am going to use chunks as an example but the same system will be used to store all data.

The current version of the core library stores the actual chunk objects in a dictionary.
The new version will store the pickled bytes representation of the chunk and unpickle it each time `get_chunk` is called. The overhead is negligible.

There will be a cache consisting of a Least Recently Used cache in RAM with a user configurable size. When that gets filled up the least recently used entries will get pushed to disk.
Users with lots of RAM can increase the size of the RAM cache for optimal speed.

When loading a chunk for the first time it will get serialised and saved to the cache. Loading from the cache is faster than loading it from the level again because it skips the translation stage.

When setting/deleting a chunk, if the history system is enabled and the original chunk state does not exist in the cache it will load the chunk from the level and store it in the cache.

Disabling the history system will reduce the amount of data generated because the cache does not have to store multiple revisions. It will also mean that the original state will not need to be loaded if it was not already loaded.

I still need to think more on how the history system will work if multiple threads are running distinct operations.
My current thought is that when an operation is finished, it will increment the undo value and any changes from then on become a new undo point. If thread A and B make changes that overlap, each will create an undo point when it is finished and the undo would revert changes made during that time period.

```
|-------Undo 1-------|----Undo 2----|
|AAAAAAAAAAAAAAAAAAAA|              |
|         BBBBBBBBBBB|BBBBBBBBBBBBBB|
```

**Comment by Rsge** [comment]
@liquidwater2 Triple? You're lucky. I am trying to replace all instances of two of a mod's stone variants with Vanilla stone in the whole Overworld (long story):
The "actual" operation takes ca. 10 mins, the `Create Undo Point` operation takes **over six hours**. I had to let my PC run over night, because I didn't know if it would continue work after being put to Sleep.
So yeah, I'd greatly second this for future use cases like mine.

**Comment by iAkariAk** [comment]
So, is the issue resolved now? I need to read a large map to extract each piece of text, but the history manager's cache causes huge overhead, so I have to divide them into many parts. Hence I need to disable it.

**Comment by gentlegiantJGC** [comment]
IIRC there is a function you can call to clear loaded chunks from memory so that they don't need to get pushed to the history manager

**Comment by iAkariAk** [comment]
> IIRC there is a function you can call to clear loaded chunks from memory so that they don't need to get pushed to the history manager

Yeah, it's `purge`. But undeniably, the function significantly slows down processing time, even though I don't need `redo` and such. Could provide an option to disable it?
